Source/Core/Core/Debugger/PPCDebugInterface.cpp

```diff
@@ -29,38 +29,41 @@
 #include "Core/PowerPC/PowerPC.h"
 #include "Core/System.h"
 
-void ApplyMemoryPatch(const Core::CPUThreadGuard& guard, Common::Debug::MemoryPatch& patch,
+void ApplyMemoryPatch(const Core::CPUThreadGuard& guard, std::span<u8> value, const u32 address,
                       bool store_existing_value)
 {
   if (AchievementManager::GetInstance().IsHardcoreModeActive())
     return;
 
-  if (patch.value.empty())
+  if (value.empty())
     return;
 
-  const u32 address = patch.address;
-  const std::size_t size = patch.value.size();
+  const std::size_t size = value.size();
   if (!PowerPC::MMU::HostIsRAMAddress(guard, address))
     return;
 
   auto& power_pc = guard.GetSystem().GetPowerPC();
+
+  bool should_invalidate_cache = false;
   for (u32 offset = 0; offset < size; ++offset)
   {
-    if (store_existing_value)
-    {
-      const u8 value = PowerPC::MMU::HostRead_U8(guard, address + offset);
-      PowerPC::MMU::HostWrite_U8(guard, patch.value[offset], address + offset);
-      patch.value[offset] = value;
-    }
-    else
+    u8 old_value = PowerPC::MMU::HostRead_U8(guard, address + offset);
+    if (old_value != value[offset])
     {
-      PowerPC::MMU::HostWrite_U8(guard, patch.value[offset], address + offset);
+      PowerPC::MMU::HostWrite_U8(guard, value[offset], address + offset);
+      should_invalidate_cache = true;
+      if (store_existing_value)
+        value[offset] = old_value;
     }
 
     if (((address + offset) % 4) == 3)
-      power_pc.ScheduleInvalidateCacheThreadSafe(Common::AlignDown(address + offset, 4));
+    {
+      if (should_invalidate_cache)
+        power_pc.ScheduleInvalidateCacheThreadSafe(Common::AlignDown(address + offset, 4));
+      should_invalidate_cache = false;
+    }
```

Comment:
I can't put a comment on it but I don't think the second invalidate call uses the right address. `address + size` puts you one byte after the end of the write, so you end up invalidating the next instruction rather than the one you wrote to.

Now... I guess with the code as written that doesn't actually matter. But it looks really weird to me because it's not immediately obvious that the 2nd invalidation can only be reached if `address + size` is not four-byte aligned.